File: src/main/kotlin/com/xs0/dbktx/DbMutationImpl.java
package com.xs0.dbktx;

import java.util.HashMap;
import java.util.LinkedHashMap;

public abstract class DbMutationImpl<E extends DbEntity<E, ID>, ID> implements DbMutation<E> {
    protected final HashMap<Column<E, ?>, Expr<? super E, ?>> values = new LinkedHashMap<>();
    protected final DbTable<E, ID> table;
    protected final DbConn<?> db;

    protected DbMutationImpl(DbConn<?> db, DbTable<E, ID> table) {
        this.table = table;
        this.db = db;
    }

    public <T> DbMutationImpl<E, ID> set(Column<E, T> column, Expr<? super E, T> value) {
        values.put(column, value);
        return this;
    }

    @Override
    public <TARGET extends DbEntity<TARGET, TID>, TID>
    DbMutation<E> set(RelToOne<E, TARGET> relation, TARGET target) {
        RelToOneImpl<E, ID, TARGET, TID> rel = (RelToOneImpl<E, ID, TARGET, TID>) relation;
        ManyToOneInfo<E, ID, TARGET, TID> info = rel.getInfo();

        for (ColumnMapping<E, TARGET, ?> colMap : info.getColumnMappings()) {
            doColMap(colMap, target);
        }

        return this;
    }

    private <TARGET extends DbEntity<TARGET, TID>, TID, VALTYPE>
    void doColMap(ColumnMapping<E, TARGET, VALTYPE> colMap, TARGET target) {
        set(colMap.getColumnFrom(), colMap.getColumnTo().from(target));
    }


}

File: src/main/kotlin/com/xs0/dbktx/DbMutationImpl.kt
package com.xs0.dbktx

abstract class DbMutationImpl<E : DbEntity<E, ID>, ID: Any> protected constructor(
        protected val db: DbConn,
        protected val table: DbTable<E, ID>) : DbMutation<E> {

    protected val values = EntityValues<E>()

    override operator fun <T: Any>
    set(column: Column<E, T>, value: Expr<in E, T>): DbMutationImpl<E, ID> {
        values.apply {
            column to value
        }
        return this
    }

    override fun <TARGET : DbEntity<TARGET, TID>, TID: Any>
    set(relation: RelToOne<E, TARGET>, target: TARGET): DbMutation<E> {
        @Suppress("UNCHECKED_CAST")
        relation as RelToOneImpl<E, ID, TARGET, TID>

        for (colMap in relation.info.columnMappings) {
            doColMap(colMap, target)
        }

        return this
    }

    private fun <TARGET : DbEntity<TARGET, TID>, TID, VALTYPE: Any>
    doColMap(colMap: ColumnMapping<E, TARGET, VALTYPE>, target: TARGET) {
        set(colMap.columnFrom, colMap.columnTo.invoke(target))
    }


}